tests/unit/archive_validate_test.py

```diff
@@ -0,0 +1,266 @@
+"""Test archive validate module."""
+import pytest
+
+from pudl_archiver.archivers import validate
+from pudl_archiver.frictionless import Resource
+
+
+@pytest.mark.parametrize(
+    "baseline_partitions,new_partitions,diffs",
+    [
+        (
+            {"part0": "val0", "part1": "val1"},
+            {"part0": "val0", "part1": "val1"},
+            [],
+        ),
+        (
+            {"part0": "val0", "part1": "val1"},
+            {"part0": "val0", "part1": "val1_changed"},
+            [
+                validate.PartitionDiff(
+                    key="part1",
+                    value="val1_changed",
+                    previous_value="val1",
+                    diff_type="UPDATE",
+                )
+            ],
+        ),
+        (
+            {"part0": "val0", "part1": "val1_deleted"},
+            {"part0": "val0"},
+            [
+                validate.PartitionDiff(
+                    key="part1",
+                    previous_value="val1_deleted",
+                    diff_type="DELETE",
+                )
+            ],
+        ),
+        (
+            {"part0": "val0"},
+            {"part0": "val0", "part1": "val1_created"},
+            [
+                validate.PartitionDiff(
+                    key="part1",
+                    value="val1_created",
+                    diff_type="CREATE",
+                )
+            ],
+        ),
+    ],
+)
+def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
+    """Test partition diffs."""
+    test_diffs = validate._process_partition_diffs(baseline_partitions, new_partitions)
+    assert test_diffs == diffs
+
+
+@pytest.mark.parametrize(
```

Comment:
Good instinct to parametrize, though I personally still find this data kind of hard to follow just because there's so many lines of it. Seems like there's a lot of repetition with small tweaks in the data - maybe consider a test data factory that churns out similar-looking data? I'll mess around a bit on Monday and see what I come up with.

Reply:
Here's an example of what such a factory might look like, and how it makes it easier to add more test cases:

```diff
diff --git a/tests/unit/archive_validate_test.py b/tests/unit/archive_validate_test.py
index 46eb31a..1b7d384 100644
--- a/tests/unit/archive_validate_test.py
+++ b/tests/unit/archive_validate_test.py
@@ -55,58 +55,30 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
     assert test_diffs == diffs
 
 
+def fake_resource(num=0, **kwargs):
+    params = {
+        "name": f"resource{num}",
+        "path": "https://www.fake.link",
+        "remote_url": "https://www.fake.link",
+        "title": f"Resource {num}",
+        "parts": {},
+        "mediatype": "zip",
+        "format": "format",
+        "bytes": 10,
+        "hash": f"hash{num}",
+    } | kwargs
+    return Resource(**params)
+
+
 @pytest.mark.parametrize(
     "baseline_resources,new_resources,diffs",
     [
         (
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1",
-                ),
-            },
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=40,
-                    hash="hash1_changed",
-                ),
-            },
+            [
+                fake_resource(0),
+                fake_resource(1, bytes=20),
+            ],
+            [fake_resource(0), fake_resource(1, bytes=40, hash="hash1_changed")],
             [
                 validate.FileDiff(
                     name="resource1",
@@ -116,43 +88,11 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
             ],
         ),
         (
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1",
-                ),
-            },
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-            },
+            [
+                fake_resource(0),
+                fake_resource(1, bytes=20),
+            ],
+            [fake_resource(0)],
             [
                 validate.FileDiff(
                     name="resource1",
@@ -162,43 +102,8 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
             ],
         ),
         (
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-            },
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1_created",
-                ),
-            },
+            [fake_resource(0)],
+            [fake_resource(0), fake_resource(1, bytes=20)],
             [
                 validate.FileDiff(
                     name="resource1",
@@ -208,59 +113,30 @@ def test_process_partition_diffs(baseline_partitions, new_partitions, diffs):
             ],
         ),
         (
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1",
-                ),
-            },
-            {
-                "resource0": Resource(
-                    name="resource0",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 0",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=10,
-                    hash="hash0",
-                ),
-                "resource1": Resource(
-                    name="resource1",
-                    path="https://www.fake.link",
-                    remote_url="https://www.fake.link",
-                    title="Resource 1",
-                    parts={},
-                    mediatype="zip",
-                    format="format",
-                    bytes=20,
-                    hash="hash1",
-                ),
-            },
+            [fake_resource(0), fake_resource(1, bytes=20)],
+            [fake_resource(0), fake_resource(1, bytes=20)],
             [],
         ),
+        (
+            [fake_resource(0)],
+            [fake_resource(0, hash="hash0_changed")],
+            [validate.FileDiff(name="resource0", diff_type="UPDATE", size_diff=0)],
+        ),
     ],
 )
def test_process_resource_diffs(baseline_resources, new_resources, diffs):
     """Test resource diffs."""
     test_diffs = validate._process_resource_diffs(baseline_resources, new_resources)
     assert test_diffs == diffs
```

Reply:
Good call, much cleaner now